File: server/routers/sse.py
import asyncio
import json
import time
from typing import Optional

from fastapi import APIRouter, Header
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel, field_validator

from infra.log import logger
from infra.telemetry.metrics import (
    chat_request_total,
    chat_request_duration,
    chat_stream_duration,
    chat_active_streams,
    _token_prefix,
)
import services.state as state
from services.bridge import CHAT_INBOX_PREFIX

router = APIRouter()

_SSE_TIMEOUT = 300  # 5 minutes
_SSE_BLOCK_MS = 1000  # XREADGROUP block timeout — short so we can send heartbeats
_HEARTBEAT_INTERVAL = 15.0  # seconds between SSE heartbeat comments


def _record_request(status: str, code: int, token_prefix: str, t0: float) -> None:
    """Record request counter + duration histogram for a given status."""
    attrs = {"status": status, "code": str(code), "token_prefix": token_prefix}
    chat_request_total.add(1, attrs)
    chat_request_duration.record(time.time() - t0, attrs)


# ---------------------------------------------------------------------------
# Request / response models
# ---------------------------------------------------------------------------

class MediaItem(BaseModel):
    type: str            # "url" | "base64" | "s3_key" (only "url" implemented now)
    content: str         # primary value: URL / base64 data / S3 key
    mimeType: str = ""   # required only for type="base64"

    @field_validator("content")
    @classmethod
    def validate_content(cls, v):
        if not v or not v.strip():
            raise ValueError("Media item content must not be empty")
        return v


class ChatRequest(BaseModel):
    content: str = ""
    sessionId: Optional[str] = None
    media: Optional[list[MediaItem]] = None

    @field_validator("media")
    @classmethod
    def validate_media(cls, v):
        if v is not None and len(v) > 10:
            raise ValueError("Too many media items (max 10)")
        return v


# ---------------------------------------------------------------------------
# Auth helper
# ---------------------------------------------------------------------------

async def _authenticate(authorization: Optional[str]) -> Optional[str]:
    """Extract and validate token from Authorization: Bearer header.

    Returns the validated token string, or None if invalid.
    """
    if not authorization or not authorization.lower().startswith("bearer "):
        return None

    token = authorization[7:].strip()
    if not token:
        return None

    if await state.token_manager.validate(token):
        return token
    return None


# ---------------------------------------------------------------------------
# Session resolution
# ---------------------------------------------------------------------------

async def _resolve_session(
    token: str,
    session_id: Optional[str],
) -> tuple[str, int]:
    """Resolve or auto-create a session.

    - If session_id is provided, validate it exists.
    - If not provided, create a new session.

    Returns (session_id, session_number).
    Raises ValueError with a message on failure.
    """
    bridge = state.bridge

    if session_id:
        match = await bridge.get_session(token, session_id)
        if not match:
            raise ValueError(f"Session not found: {session_id}")
        return match[0], match[1]

    # Create new session
    return await bridge.create_session(token)


# ---------------------------------------------------------------------------
# SSE formatting
# ---------------------------------------------------------------------------

def _sse_event(event_type: str, data: dict) -> str:
    return f"event: {event_type}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"


def _sse_comment() -> str:
    return ": heartbeat\n\n"


# ---------------------------------------------------------------------------
# SSE stream generator
# ---------------------------------------------------------------------------

async def _stream_response(
    token: str,
    session_id: str,
    session_number: int,
    req_id: str,
):
    """Consume events from Redis Stream inbox and yield SSE events."""
    queue = state.queue
    inbox = f"{CHAT_INBOX_PREFIX}{token}:{session_id}"
    deadline = time.time() + _SSE_TIMEOUT
    last_heartbeat = time.time()

    # First event: session info
    yield _sse_event("session", {
        "sessionId": session_id,
        "sessionNumber": session_number,
    })

    try:
        while time.time() < deadline:
            result = await queue.consume(
                inbox, group="sse", consumer=req_id,
                block_ms=_SSE_BLOCK_MS,
            )

            if result is None:
                # No message — send heartbeat if interval elapsed
                now = time.time()
                if now - last_heartbeat >= _HEARTBEAT_INTERVAL:
                    yield _sse_comment()
                    last_heartbeat = now
                else:
                    await asyncio.sleep(0.1)
                continue

            msg_id, raw = result
            await queue.ack(inbox, "sse", msg_id)
            await queue.delete_message(inbox, msg_id)

            try:
                event = json.loads(raw)
            except json.JSONDecodeError:
                logger.warning("SSE: invalid JSON in inbox (token={}...)", token[:10])
                continue

            event_type = event.pop("type", "message")

            # For media events, the payload is nested under "data" key
            # so that SSE data contains only {type, content} without the event type
            if event_type == "media":
                event_data = event.pop("data", None)
                if not event_data:
                    logger.warning("SSE: media event missing data payload (token={}...)", token[:10])
                    continue
            else:
                event_data = event
            yield _sse_event(event_type, event_data)

            # Terminal events — close the stream
            if event_type in ("done", "error"):
                return

        # Timeout reached
        yield _sse_event("error", {"content": "Stream timeout"})
    except asyncio.CancelledError:
        # Client disconnected
        logger.info("SSE: client disconnected (token={}...)", token[:10])
    except Exception:
        logger.exception("SSE: stream error (token={}...)", token[:10])
        yield _sse_event("error", {"content": "Internal server error"})


async def _stream_with_cleanup(
    token: str,
    session_id: str,
    session_number: int,
    req_id: str,
):
    """Wrap _stream_response and delete the chat inbox stream when done."""
    tp = _token_prefix(token)
    stream_attrs = {"token_prefix": tp}

    # Track active stream
    chat_active_streams.add(1, stream_attrs)
    stream_start = time.time()
    close_reason = "done"

    try:
        async for event in _stream_response(token, session_id, session_number, req_id):
            yield event
            # Detect close reason from terminal events
            if event.startswith("event: error"):
                if "Stream timeout" in event:
                    close_reason = "timeout"
                else:
                    close_reason = "error"
    except asyncio.CancelledError:
        close_reason = "client_disconnect"
        raise
    except Exception:
        close_reason = "error"
        raise
    finally:
        chat_active_streams.add(-1, stream_attrs)
        stream_duration = time.time() - stream_start

        chat_stream_duration.record(
            stream_duration,
            {"close_reason": close_reason, "token_prefix": tp},
        )

        try:
            inbox = f"{CHAT_INBOX_PREFIX}{token}:{session_id}"
            await state.queue.delete_queue(inbox)
        except Exception:
            logger.warning("SSE: cleanup failed (token={}...)", token[:10])


# ---------------------------------------------------------------------------
# POST /bridge/chat — Dialogue endpoint (SSE stream response)
# ---------------------------------------------------------------------------

@router.post("/bridge/chat")
async def chat_sse(
    body: ChatRequest,
    authorization: Optional[str] = Header(default=None),
):
    t0 = time.time()

    # Authenticate
    token = await _authenticate(authorization)
    if not token:
        logger.warning("SSE: auth failed (token missing or invalid)")
        _record_request("auth_fail", 401, "", t0)
        return JSONResponse(
            status_code=401,
            content={"ok": False, "error": "Invalid or missing token"},
        )

    tp = _token_prefix(token)

    # Validate message content — normalize media URLs
    content = body.content or ""
    media_urls: list[str] = []

    if body.media:
        for item in body.media:
            if item.type == "url":
                if not item.content.startswith(("http://", "https://")):
                    logger.warning("SSE: bad request — invalid media URL scheme: {} (token={}...)", item.content, token[:10])
                    _record_request("bad_request", 400, tp, t0)
                    return JSONResponse(
                        status_code=400,
                        content={"ok": False, "error": f"Invalid media URL scheme: {item.content}"},
                    )
                media_urls.append(item.content)
            else:
                logger.warning("SSE: bad request — unsupported media type: {} (token={}...)", item.type, token[:10])
                _record_request("bad_request", 400, tp, t0)
                return JSONResponse(
                    status_code=400,
                    content={"ok": False, "error": f"Unsupported media type: {item.type}"},
                )

    if not content and not media_urls:
        logger.warning("SSE: bad request — empty message (token={}...)", token[:10])
        _record_request("bad_request", 400, tp, t0)
        return JSONResponse(
            status_code=400,
            content={"ok": False, "error": "Empty message"},
        )

    # Check bot connected
    if not await state.bridge.is_bot_connected(token):
        logger.warning("SSE: no bot connected (token={}...)", token[:10])
        _record_request("no_bot", 400, tp, t0)
        return JSONResponse(
            status_code=400,
            content={"ok": False, "error": "No bot connected"},
        )

    # Resolve session
    try:
        session_id, session_number = await _resolve_session(token, body.sessionId)
    except ValueError as e:
        logger.warning("SSE: session not found {} (token={}...)", body.sessionId, token[:10])
        _record_request("session_not_found", 404, tp, t0)
        return JSONResponse(
            status_code=404,
            content={"ok": False, "error": str(e)},
        )

    # Clear stale events and reset consumer group for this SSE request
    queue = state.queue
    inbox = f"{CHAT_INBOX_PREFIX}{token}:{session_id}"
    await queue.purge(inbox)
    await queue.ensure_group(inbox, "sse")

    # Send message to bot via Redis Stream inbox
    req_id = await state.bridge.send_to_bot(
        token, content,
        media_urls=media_urls or None,
        session_id=session_id,
    )
    if not req_id:
        logger.error("SSE: send_to_bot failed (token={}...)", token[:10])
        _record_request("send_fail", 500, tp, t0)
        return JSONResponse(
            status_code=500,
            content={"ok": False, "error": "Failed to send message to bot"},
        )

    # Success — entering SSE stream
    _record_request("success", 200, tp, t0)

    logger.info(
        "SSE: chat started req={} session={} (token={}...)",
        req_id, session_id[:8], token[:10],
    )

    return StreamingResponse(
        _stream_with_cleanup(token, session_id, session_number, req_id),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


# ---------------------------------------------------------------------------
# GET /bridge/chat/sessions — List sessions
# ---------------------------------------------------------------------------

@router.get("/bridge/chat/sessions")
async def list_sessions(
    authorization: Optional[str] = Header(default=None),
):
    validated = await _authenticate(authorization)
    if not validated:
        logger.warning("SSE: sessions auth failed (list)")
        return JSONResponse(
            status_code=401,
            content={"ok": False, "error": "Invalid or missing token"},
        )

    sessions = await state.bridge.get_sessions(validated)

    return {
        "ok": True,
        "sessions": [{"id": s[0], "number": s[1]} for s in sessions],
    }


# ---------------------------------------------------------------------------
# POST /bridge/chat/sessions — Create new session
# ---------------------------------------------------------------------------

@router.post("/bridge/chat/sessions")
async def create_session(
    authorization: Optional[str] = Header(default=None),
):
    validated = await _authenticate(authorization)
    if not validated:
        logger.warning("SSE: sessions auth failed (create)")
        return JSONResponse(
            status_code=401,
            content={"ok": False, "error": "Invalid or missing token"},
        )

    session_id, session_number = await state.bridge.create_session(validated)
    sessions = await state.bridge.get_sessions(validated)

    return {
        "ok": True,
        "sessionId": session_id,
        "sessionNumber": session_number,
        "sessions": [{"id": s[0], "number": s[1]} for s in sessions],
    }

File: backend/internal/router/sse.go
package router

import (
	"context"
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/rs/zerolog/log"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"

	"astron-claw/backend/internal/infra/telemetry"
	"astron-claw/backend/internal/model"
	"astron-claw/backend/internal/pkg"
	"astron-claw/backend/internal/service"
)

const (
	sseTimeout        = 600 // 10 minutes
	sseBlockMs        = 1000
	heartbeatInterval = 15.0 // seconds
)

type ChatRequest struct {
	Content   string      `json:"content"`
	SessionID *string     `json:"sessionId,omitempty"`
	Media     []MediaItem `json:"media,omitempty"`
}

type MediaItem struct {
	Type     string `json:"type"`
	Content  string `json:"content"`
	MimeType string `json:"mimeType,omitempty"`
}

func (app *App) chatSSE(c *gin.Context) {
	t0 := time.Now()
	tokenStr := c.GetString("token")
	tp := telemetry.TokenPrefix(tokenStr)

	reqStatus := "success"
	reqCode := 200
	recordReq := func() {
		attrs := metric.WithAttributeSet(attribute.NewSet(
			attribute.String("status", reqStatus),
			attribute.String("code", strconv.Itoa(reqCode)),
			attribute.String("token_prefix", tp),
		))
		telemetry.ChatRequestTotal.Add(context.Background(), 1, attrs)
		telemetry.ChatRequestDuration.Record(context.Background(), time.Since(t0).Seconds(), attrs)
	}
	defer recordReq()

	var body ChatRequest
	if err := c.ShouldBindJSON(&body); err != nil {
		reqStatus = "bad_request"
		reqCode = 400
		model.ErrorResponse(c, model.ErrChatInvalidReq)
		return
	}

	// Validate media
	content := body.Content
	var mediaURLs []string

	if len(body.Media) > 10 {
		log.Warn().Str("token", tp).Msg("SSE: bad request — too many media items")
		reqStatus = "bad_request"
		reqCode = 400
		model.ErrorResponse(c, model.ErrMediaTooMany)
		return
	}

	if body.Media != nil {
		for _, item := range body.Media {
			if item.Type == "url" {
				if !strings.HasPrefix(item.Content, "http://") && !strings.HasPrefix(item.Content, "https://") {
					log.Warn().Str("url", item.Content).Str("token", tp).
						Msg("SSE: bad request — invalid media URL scheme")
					reqStatus = "bad_request"
					reqCode = 400
					model.ErrorResponse(c, model.ErrMediaBadURLScheme)
					return
				}
				mediaURLs = append(mediaURLs, item.Content)
			} else {
				log.Warn().Str("type", item.Type).Str("token", tp).
					Msg("SSE: bad request — unsupported media type")
				reqStatus = "bad_request"
				reqCode = 400
				model.ErrorResponse(c, model.ErrMediaUnsupportedType)
				return
			}
		}
	}

	if content == "" && len(mediaURLs) == 0 {
		log.Warn().Str("token", tp).Msg("SSE: bad request — empty message")
		reqStatus = "bad_request"
		reqCode = 400
		model.ErrorResponse(c, model.ErrChatEmptyMessage)
		return
	}

	// Check bot connected
	ctx := c.Request.Context()
	if !app.Bridge.IsBotConnected(ctx, tokenStr) {
		log.Warn().Str("token", tp).Msg("SSE: no bot connected")
		reqStatus = "no_bot"
		reqCode = 400
		model.ErrorResponse(c, model.ErrChatNoBot)
		return
	}

	// Resolve session
	var sessionID string
	var sessionNumber int
	if body.SessionID != nil && *body.SessionID != "" {
		sid, snum, found := app.Bridge.GetSession(ctx, tokenStr, *body.SessionID)
		if !found {
			log.Warn().Str("session", *body.SessionID).Str("token", tp).
				Msg("SSE: session not found")
			reqStatus = "session_not_found"
			reqCode = 404
			model.ErrorResponse(c, model.ErrSessionNotFound)
			return
		}
		sessionID = sid
		sessionNumber = snum
	} else {
		var err error
		sessionID, sessionNumber, err = app.Bridge.CreateSession(ctx, tokenStr)
		if err != nil {
			log.Error().Err(err).Str("token", tp).Msg("SSE: failed to create session")
			reqStatus = "error"
			reqCode = 500
			c.JSON(500, gin.H{"code": 500, "error": "Failed to create session"})
			return
		}
	}

	// Clear stale events and reset consumer group
	inbox := service.ChatInboxPrefix + tokenStr + ":" + sessionID
	app.Queue.Purge(ctx, inbox)
	app.Queue.EnsureGroup(ctx, inbox, "sse")
	app.Bridge.TrackChatInbox(ctx, tokenStr, inbox)

	// Send message to bot
	reqID, err := app.Bridge.SendToBot(ctx, tokenStr, content, mediaURLs, sessionID)
	if err != nil {
		log.Error().Err(err).Str("token", tp).Msg("SSE: send_to_bot failed")
		reqStatus = "send_fail"
		reqCode = 500
		model.ErrorResponse(c, model.ErrChatSendFailed)
		return
	}

	// Success — entering SSE stream

	log.Info().Str("req", reqID).Str("session", pkg.SafePrefix(sessionID, 8)).Str("token", tp).
		Msg("SSE: chat started")

	flusher, ok := c.Writer.(http.Flusher)
	if !ok {
		c.JSON(500, gin.H{"code": 500, "error": "Streaming not supported"})
		return
	}

	// Set SSE headers
	c.Header("Content-Type", "text/event-stream")
	c.Header("Cache-Control", "no-cache")
	c.Header("Connection", "keep-alive")
	c.Header("X-Accel-Buffering", "no")
	c.Status(http.StatusOK)

	// Track active stream
	streamStart := time.Now()
	closeReason := "done"
	telemetry.ChatActiveStreams.Add(ctx, 1)

	defer func() {
		telemetry.ChatActiveStreams.Add(context.Background(), -1)
		streamDuration := time.Since(streamStart).Seconds()
		telemetry.ChatStreamDuration.Record(context.Background(), streamDuration,
			metric.WithAttributeSet(attribute.NewSet(
				attribute.String("close_reason", closeReason),
				attribute.String("token_prefix", tp),
			)),
		)
	}()

	// Stream events — use a deadline tracker instead of re-creating contexts
	deadline := time.Now().Add(sseTimeout * time.Second)

	// First event: session info
	sessionEvent := pkg.FormatSSEEvent("session", map[string]interface{}{
		"sessionId":     sessionID,
		"sessionNumber": sessionNumber,
	})
	if _, err := c.Writer.WriteString(sessionEvent); err != nil {
		closeReason = "write_error"
		return
	}
	flusher.Flush()

	lastHeartbeat := time.Now()
	hasChunks := false

	for {
		// Check client disconnect
		select {
		case <-c.Request.Context().Done():
			closeReason = "client_disconnect"
			log.Info().Str("token", tp).Msg("SSE: client disconnected")
			return
		default:
		}

		// Check timeout
		if time.Now().After(deadline) {
			closeReason = "timeout"
			errEvent := pkg.FormatSSEEvent("error", map[string]interface{}{
				"content": model.ErrChatStreamTimeout.Message,
			})
			_, _ = c.Writer.WriteString(errEvent)
			flusher.Flush()
			return
		}

		result, err := app.Queue.Consume(c.Request.Context(), inbox, "sse", reqID, sseBlockMs)
		if err != nil {
			log.Error().Err(err).Str("token", tp).Msg("SSE: consume error")
			closeReason = "error"
			errEvent := pkg.FormatSSEEvent("error", map[string]interface{}{
				"content": model.ErrChatInternalError.Message,
			})
			_, _ = c.Writer.WriteString(errEvent)
			flusher.Flush()
			return
		}

		if result == nil {
			// Check client disconnect
			select {
			case <-c.Request.Context().Done():
				closeReason = "client_disconnect"
				return
			default:
			}

			// Heartbeat
			if time.Since(lastHeartbeat).Seconds() >= heartbeatInterval {
				_, _ = c.Writer.WriteString(pkg.FormatSSEComment())
				flusher.Flush()
				lastHeartbeat = time.Now()
			}
			continue
		}

		_ = app.Queue.Ack(context.Background(), inbox, "sse", result.ID)

		// Reset deadline on activity
		deadline = time.Now().Add(sseTimeout * time.Second)

		var event map[string]interface{}
		if err := json.Unmarshal([]byte(result.Data), &event); err != nil {
			log.Warn().Str("token", tp).Msg("SSE: invalid JSON in inbox")
			continue
		}

		eventType, _ := event["type"].(string)
		if eventType == "" {
			eventType = "message"
		}
		delete(event, "type")

		var eventData map[string]interface{}
		if eventType == "media" {
			if d, ok := event["data"].(map[string]interface{}); ok {
				eventData = d
			} else {
				log.Warn().Str("token", tp).Msg("SSE: media event missing data payload")
				continue
			}
		} else {
			eventData = event
		}

		if eventType == "chunk" {
			hasChunks = true
		}

		// Auto-inject chunk before done if no preceding chunks
		if eventType == "done" && !hasChunks {
			if contentStr, ok := eventData["content"].(string); ok && contentStr != "" {
				chunkEvent := pkg.FormatSSEEvent("chunk", map[string]interface{}{
					"content": contentStr,
				})
				_, _ = c.Writer.WriteString(chunkEvent)
				flusher.Flush()
			}
		}

		sseEvent := pkg.FormatSSEEvent(eventType, eventData)
		_, _ = c.Writer.WriteString(sseEvent)
		flusher.Flush()

		// Terminal events
		if eventType == "done" || eventType == "error" {
			return
		}
	}
}

func (app *App) listSessions(c *gin.Context) {
	tokenStr := c.GetString("token")

	sessions, err := app.Bridge.GetSessions(c.Request.Context(), tokenStr)
	if err != nil {
		log.Error().Err(err).Msg("Failed to list sessions")
		c.JSON(500, gin.H{"code": 500, "error": "Internal server error"})
		return
	}

	sessionList := make([]gin.H, len(sessions))
	for i, s := range sessions {
		sessionList[i] = gin.H{"id": s.ID, "number": s.Number}
	}

	c.JSON(200, gin.H{
		"code":     0,
		"sessions": sessionList,
	})
}

func (app *App) createSession(c *gin.Context) {
	tokenStr := c.GetString("token")
	ctx := c.Request.Context()

	sessionID, sessionNumber, err := app.Bridge.CreateSession(ctx, tokenStr)
	if err != nil {
		log.Error().Err(err).Msg("Failed to create session")
		c.JSON(500, gin.H{"code": 500, "error": "Internal server error"})
		return
	}

	sessions, err := app.Bridge.GetSessions(ctx, tokenStr)
	if err != nil {
		log.Error().Err(err).Msg("Failed to list sessions")
		c.JSON(500, gin.H{"code": 500, "error": "Internal server error"})
		return
	}

	sessionList := make([]gin.H, len(sessions))
	for i, s := range sessions {
		sessionList[i] = gin.H{"id": s.ID, "number": s.Number}
	}

	c.JSON(200, gin.H{
		"code":          0,
		"sessionId":     sessionID,
		"sessionNumber": sessionNumber,
		"sessions":      sessionList,
	})
}
